UI: New ACLs

johncowen opened the pull request:
~~This is potentially a temporary PR for ease of early review.~~

~~The description of this PR is likely to be edited/added to next week and the base branch is also likely to change.~~

This PR adds new style ACLs to the UI.

Generally the new code follows the same style as the rest of the app.

There are some new concepts/additions that I've tried to keep purely in the new ACLs areas for the moment, but I plan on moving these across into the rest of the app all being well.

I'll briefly list these additions here (to be expanded upon)

## Forms

Forms are view layer objects that are mainly concerned with 'change' events of HTML form elements in a 'data down, events up' way. The allow you to define reusable forms that can be composed into larger forms. In my current case I needed to do this for embedding a policy form into a token form.

For example, creating policies is now part of editing a token, but also policies can be created outside of tokens.

I didn't want to have to repeat the set up of the validations and the 'change' listeners in both places. An alternative to 'forms' would have been writing a component for each form, but then I probably would have needed 2 components for each type of policy form, or lots of conditionals within one. This is also part one of doing something I've been wanting to do which is similar to how the vault UI dynamically generates its form views depending on the data/model.

Lastly, 'forms' integrate well with `ember-changeset-validations` and follow some conventions there.

## dom-buffer

Usecase: I want to buffer a DOM element (or HTMLBars string) from being inserted into the DOM and then flush it out elsewhere, later on.

For modals you generally have to append your modal DOM elsewhere in the DOM. I initially looked at `ember-wormhole` but due to its use of private API's and the way you use it not being very ember-y, I also looked at `ember-elsewhere` (which has similar reasons for not using `ember-wormhole`). I couldn't see if `ember-elsewhere` could be used without passing you DOM to it as an attribute. So in the end it was easier to write my own simpler thing.

Other things to note:
1. You can extend the dom-buffer component (as I do here) and tell your component to appear somewhere else.
2. Its based on injectable services.
3. It's still a work in progress, but only the straightforward things are left to do, for the moment there is only 1 buffer called 'modal'.

I also looked at `ember-maybe-element`, I need to look at this further, but at first glance it also uses private APIs currently, and also seems to be a processor rather than anything else (they also mention not wanting to use ember-wormhole)

## Dedicated `dom` service

I decided to move all of my dom utilities into a service. This means that ones that need setting up, only need setting up once, in the service itself, and they are ready to be used and injected where-ever you need them. This feels very much like a jQuery like way to do things, so I'm still slightly uncomfortable doing this, but it means I can inject and unit test things properly without resorting to things like `js-dom` or `min-document`. Moving all these to a service means that injecting all the things in one place  is less painful than in multiple places.

Lastly here, I'm still deliberating on whether to names of methods here, and I still need to finish up migrating everything to use the service instead of importing. 

## `repo.status()`

The way I was catching errors previously never transitioned to the URL you had clicked before displaying the error page. I think this is just the way the ember catchall error catching works? I needed someway to easily display errors whilst making sure the URL was the correct URL, so I'm moving away form the ember catchall error and setting `isEnabled` and `isAuthorized` properties for each route depending on the HTTP status instead. `repo.status()` is a nice way of doing this that works in with my `app-view` to show error pages whilst maintaining the correct url. 

## SingleRoute

This is my first attempt at reducing the repetition of create/edit routes. I'm still not massively convinced by this myself, and whilst is good enough for what I wanted to do at first, I think I'd prefer not extending something and using a utility function within `model` instead. This would keep the traditional `model` method a lot clearer. Completely in two minds over this still!

Lastly, this entirely PR still needs a good spring clean, there are still a lot of notes in the comments for my own use, and this won't be merged without a full spring clean and possibly a second review.

## New localStorage format

We briefly chatted about this offline. I'm now following vaults format for localStorage keys i.e. 'consul:keyname', all values are JSON encoded, even strings. I did mention that I didn't want to save JSON blobs, but I should qualify that now by saying I don't want to save big JSON blobs. Right now I'm saving little JSON blobs where it makes sense (in this case your new type of token - `{"AccessorID": "", "SecretID": ""}`, feels like it makes sense to just save that as a blob.

Important thing to note here is that there is nothing here to 'upgrade' from the old type of localStorage tokens to the new ones. I'd also rather people relogged in so they get a proper full token with ID and secret.

johncowen commented on ui-v2/app/adapters/policy.js:
```diff
@@ -0,0 +1,75 @@
+          break;
+        default:
+          response = this.handleSingleResponse(url, response, PRIMARY_KEY, SLUG_KEY);
+      }
```
The new ACLs return arrays for multiple items and objects for singular. This is much easier to figure what the response is for an add the unique id's appropriately. (don't know if you remember the `isQueryRecord` stuff from the rest of the app?)

johncowen commented on ui-v2/app/models/service.js:
```diff
@@ -21,6 +21,7 @@ export default Model.extend({
+  // TODO: These should be typed
   ChecksPassing: attr(),
   ChecksCritical: attr(),
   ChecksWarning: attr(),
```
I'm trying to avoid making any non-ACL changes in this PR, I'll probably hit this in my spring clean before final merge.

johncowen commented on ui-v2/app/routes/dc/acls/tokens/index.js:
```diff
@@ -0,0 +1,29 @@
+    return hash({
+      ...repo.status({
+        items: repo.findAllByDatacenter(this.modelFor('dc').dc.Name),
+      }),
```
Here's one place where `repo.status()` is being used. The idea is it doesn't move too far away from what I was doing before. The `status` method basically catches any HTTP errors (in this case 401 and 403) and then resolves with extra properties on the `{item: ...}` so I know whether ACLs are enabled or whether you are authorised.

The `repo` seemed like the best place to do this, still about 80% on that decision though.

johncowen replied:
Looking at the new API missing things give me 403's, but yeah good call, will check this.

johncowen replied:
Just tried this, changed it so 404's (which are actually 403's) to give me the ACL login, thanks for that!

meirish commented on ui-v2/app/adapters/policy.js:
```diff
@@ -0,0 +1,91 @@
+import { PUT as HTTP_PUT } from 'consul-ui/utils/http/method';
+
+export default Adapter.extend({
+  // _ajaxRequest(options) {
```
Imagine you don't still need this :)

meirish commented on ui-v2/app/templates/dc/acls/policies/-notifications.hbs:
```diff
@@ -0,0 +1,20 @@
+{{#if (eq type 'create')}}
```
Though the if / else party here is a bit iffy (pun intended :D) - this will be nice when we have to translate the UIs.

johncowen replied:
Oh I know, it's lovely isn't it 😄 . I don't think there are any `switch`es in `htmlbars` - I've been trying to keep all the text content in the templates, although I'm mulling that over again 🤔

meirish commented on ui-v2/app/templates/dc/acls/tokens/index.hbs:
```diff
@@ -0,0 +1,134 @@
+            {{/block-slot}}
+            {{#block-slot 'actions' as |index change checked|}}
+                {{#confirmation-dialog confirming=false index=index message="Are you sure you want to delete this Token?"}}
+                    {{#block-slot 'action' as |confirm|}}
```
Block inception 💥 💥

meirish commented on ui-v2/app/utils/acls-status.js:
```diff
@@ -0,0 +1,48 @@
+        };
+      }),
+      [propName]: p
+        .catch(function(e) {
```
Took me a minute to follow these through, but this is a clever way of wrapping the call.

meirish commented on ui-v2/app/adapters/policy.js:
```diff
@@ -0,0 +1,91 @@
+  },
+  urlForCreateRecord: function(modelName, snapshot) {
+    return this.appendURL('acl/policy', [], {
+      [API_DATACENTER_KEY]: snapshot.attr(DATACENTER_KEY),
```
follow up - aha this is mostly for dc stuff

johncowen replied:
Yeah this is just `?dc=dc-1`

meirish commented on ui-v2/app/controllers/dc/acls/policies/edit.js:
```diff
@@ -0,0 +1,38 @@
+            throw e;
+        }
+      }
+    },
```
Just curious why you didn't go with the model -> component -> service -> model with this?  I think the current implementation is roughly model -> controller -> service -> dom -> service -> model - and for it to work you have to remember a certain naming system for the inputs (though the addition of throwing would help - I think currently it doesn't? or at least there was a comment that said "we should throw here" :) ). Components would also let you do view-only attributes, and give you hooks to set up the forms instead of doing the `setProperties` override (which I felt was pretty surprising).

meirish replied:
Woke up with another thought about this 😛 - why not use the action to pass the item and prop name in? Then you wouldn't have to fish around in the DOM and name match from field attrs.

johncowen replied:
Hey, sorry a bit postscript this..

So I had a good experience this week because I was using the standard `name=""` attribute rather than passing the name through as an action argument.

The `fillIn` method in my testing 'just works' if you use `name=""`, not only that but this week whilst I was bringing the old ACL stuff (inc. its testing) back into this, I'd noticed when I came to add/tweak the testing back in that as I was reusing the new 'login' form I needed to change the fillIn value of the old tests to use the new form. I couldn't remember what property I'd used in the new form, but instead of delving into my ember code to find it, I just did an 'Inspect' and there it was in the `name="name of property"` attribute in the element inspector. Probably saved me a few seconds, but a nice detail I got for free by following a standard.

johncowen commented on ui-v2/app/adapters/policy.js:
```diff
@@ -0,0 +1,91 @@
+  urlForTranslateRecord: function(modelName, snapshot) {
+    return this.appendURL('acl/policy/translate', [], {});
+  },
+  dataForRequest: function(params) {
```
Yeah for sure, this just follows the rest of the app and the old ACLs system.

`_super` calls this in RESTAdapter

https://github.com/emberjs/data/blob/e20597e8b1a2a6d7a4882ff91cdbe7481bee9415/addon/adapters/rest.js#L1261

> RESTAdapter#ajax has been deprecated please use. `methodForRequest`, `urlForRequest`, `headersForRequest` or `dataForRequest` instead.

https://github.com/emberjs/data/blob/e20597e8b1a2a6d7a4882ff91cdbe7481bee9415/addon/adapters/rest.js#L1230

🤷‍♂️ 

I started a refactor to move this around a bit to move some of this into the serializer instead, I did all the tests so I could do that with confidence - just not been able to do the moving yet.

As this PR isn't doing that, it would be the wrong time to do it in this PR. Once this is out, and it is a good time, I'll continue the work to move this around slightly.

I'm half waiting to move to ember/ember-data 3.4 to be honest, I plan on jigging around most of the ember-data layer once that's done, I'd probably say the ember-data code is weirdly the most involved code in the app 🤷‍♂️ 

Remember all of this:

> > RESTAdapter#ajax has been deprecated please use. `methodForRequest`, `urlForRequest`, `headersForRequest` or `dataForRequest` instead.

Is what they took out of ember-data in 3.whatever it was

Triple 🤷‍♂️

meirish replied:
Ohhh that's the feature flag they moved into an addon - odd that that duplicates serializer stuff? 🤷‍♂️ for sure haha

johncowen commented on ui-v2/app/adapters/policy.js:
```diff
@@ -0,0 +1,91 @@
+    let response = payload;
+    if (status === HTTP_OK) {
+      const url = this.parseURL(requestData.url);
+      switch (true) {
```
Yeah it follows the rest of the app, this one is simpler but the others in the rest of the app aren't so simple. I'd rather follow the same pattern and be consistent. 'Generally' I'll use a `switch` instead of lots of `if..else if`s whereas if its just an `if ...else` I'll generally keep with an `if`.

johncowen commented on ui-v2/app/components/modal-dialog.js:
```diff
@@ -0,0 +1,87 @@
+      if (this.element) {
+        const win = [...get(this, 'dom').element('[role="dialog"] > div > div', this.element)][0];
+        const rect = win.getBoundingClientRect();
+        set(this, 'window', win);
```
Ah, wondered what you meant then, yeah its the modal window, maybe `dialog` would be a better name? Or `panel`, basically its the windowy bit of the modal (so not the semi-transparent fullscreen cover). Might go with `dialog` - any other suggestions appreciated!

meirish replied:
Ah nice - either of those would be great!

johncowen commented on ui-v2/app/components/modal-dialog.js:
```diff
@@ -0,0 +1,87 @@
+  didReceiveAttrs: function() {
+    this._super(...arguments);
+    if (get(this, 'name')) {
+      set(this, 'checked', false);
```
`checked` is whether the modal is visible or not. Actually this might be left over from me noodling with the modals, it's doesn't make sense that just because it has a name it means it's not visible. Will recheck this in a bit 👍

johncowen commented on ui-v2/app/components/modal-dialog.js:
```diff
@@ -0,0 +1,87 @@
+      }
+    },
+    close: function() {
+      document.getElementById('modal_close').checked = true;
```
Yeah, the `modal-layer` (the place where modals are displayed) as a radio button, if this is checked you can guarantee there are no modals visible, with CSS so its guaranteed not to break. I would probably say its a global-ish 'setting' yeah. This is one of the places what I want to replace with my `service('dom')`.

johncowen commented on ui-v2/app/controllers/dc/acls/policies/edit.js:
```diff
@@ -0,0 +1,38 @@
+    change: function(e, value, item) {
+      try {
+        get(this, 'form').handleEvent(get(this, 'dom').normalizeEvent(e, value));
+      } catch (e) {
```
Hmm yeah good point, 2 `e`s, not nice! Will change! Actually looking at the surrounding code I wonder why I used the error and didn't just reference the original normalised event? Will have another review of this myself.

johncowen replied:
Good spot this one, reads clearer now

johncowen commented on ui-v2/app/controllers/dc/acls/policies/edit.js:
```diff
@@ -0,0 +1,38 @@
+  builder: service('form'),
+  dom: service('dom'),
+  isScoped: false,
+  setProperties: function(model) {
```
> Overriding setProperties like this is bad imo

It's a public method? Could you explain why you think it's bad? Again I'm already doing this in other places in the app (including the old ACLs that are being replaced by this)

I don't think `setupController` fits my needs here, I know of at least one place where I am calling `setProperties` from the route after controller setup, so if I moved it to `setupController` as far as I understand it my extra code then wouldn't be called?

meirish replied:
Just the principle of least surprise - `setProperties` is usually for setting a hash on an ember object. I found it surprising that suddenly it has side effects and that for things like forms to work you had to call `setProperties`.

johncowen replied:
Ah getcha, yeah I suppose. Here as the forms basically wrap the hash you send in with a `ember-validation-changeset` which is just a proxy object to the hash you've sent.

You don't 'have' to do this in `setProperties`, you could do it anywhere where you have the properties available. I just know everywhere in the app the first time the hash is set is in `setProperties`, then I know that whenever I mutate those I always use `setProperties` to do it.

We can chat more tomorrow, this is one of the things it would be good to chat through with you.

johncowen commented on ui-v2/app/controllers/dc/acls/tokens/edit.js:
```diff
@@ -0,0 +1,72 @@
+      }
+      get(this, 'dom')
+        .component(selector, parent)
+        .didAppear();
```
Ah this was brought up previously here:

https://github.com/hashicorp/consul/pull/4190#discussion_r193270223

Same comments apply here, plus its what I'm already doing in the rest of the app, so I want to stay consistent.

I did spend quite a lot of time looking into the best way to do this within ember, most ember-y people out there do similar things (i.e. don't use what you would call traditional ember methods).

Re: data down actions up, it is being called on an event listener, it just calls a method instead of passing a property back down. Really no state is changing so that kind of make sense. It's only needed because when codemirror is initially rendered it has a `display :none` so it doesn't render properly on 'didInsertElement` this was just my 'ember-y' sounding lifecycle hook for when the element is actually displayed.

johncowen replied:
Hey,

So the example you've given here is exactly what I want to avoid.

We've spoken about this before in a previous PR and offline following that, I'm not sure if you remember but I shared a few links with you, one was from Sam Selikoff (who I believe maintains mirage?)

http://www.samselikoff.com/blog/getting-ember-components-to-respond-to-actions/

> Using data-binding we could introduce a dummy property, have the street view component change that property, and have the fullscreen component respond to changes on that property; but this feels hackish, and the dummy variable wouldn't really represent anything meaningful in our application.

https://emberigniter.com/communication-between-distant-components/

This one ^ isn't _that_ similar to what I want to do here, and even this EventBus example is too coupled for me for my specific case, but:

> On the other hand, “Data down actions up” ain’t no religion. It’s important to know our tools.

I did see another article recently (dated a few weeks ago) that also discussed the same thing and had a similar solution, if I can find it again I'll post it here. I'll see if I can find that old PR were we discussed this also.

johncowen replied:
Hey,

I found the PR where we spoke about this previously:

https://github.com/hashicorp/consul/pull/4190#discussion_r193270223

I also found the more recent article related to this last night:

https://medium.com/freshworks-engineering-blog/ember-accessing-child-component-functions-in-parent-f7b0fc1ad7fa

^ This one was featured in the Ember Times 3-4 weeks ago.

John

johncowen commented on ui-v2/app/mixins/creating-route.js:
```diff
@@ -0,0 +1,23 @@
+    // Change changeset to support ember-data props
+    const item = get(this.controller, 'item.data');
+    if (get(item, 'isNew')) {
+      item.destroyRecord();
```
I think we went through this back when the consul v2 rebuild first went out. ember-data leaves unsaved objects hanging around, so if you don't unload them empty ones can end up in your lists. I'm pretty sure `rollbackAttributes` just resets the object, it doesn't completely wipe it from ember-data's cache?

This is another bit that I've just used from whats happening in the app previously, I wouldn't want to change it now for no reason and introduce any errors.

meirish replied:
If the object isNew, ember-data unloads the model from the store when you call `rollbackAttributes` - just brought it up since you were checking isNew already.

johncowen replied:
I've added this to my todo to look at changing 🔜 (right now I use it everywhere else in the app so I don't want to tweak in this PR)

johncowen commented on ui-v2/app/mixins/token/with-actions.js:
```diff
@@ -0,0 +1,60 @@
+                return this.afterDelete(...arguments);
+              });
+          });
+      }, 'use');
```
It's the name of the action you are performing. They eventually end up as like an `notification.type` in the template.

johncowen commented on ui-v2/app/mixins/token/with-actions.js:
```diff
@@ -0,0 +1,60 @@
+                // using is similar to delete in that
+                // if you use from the listing page, stay on the listing page
+                // whereas if you use from the detail page, take me back to the listing page
+                return this.afterDelete(...arguments);
```
😄 hehe yeah arrow functions 🤷‍♂️ I try to use them sparingly.

You're right tho, in this case it could be confusing for someone looking at this. The code is kind of a 'pipeline' so `arguments` is always the same `item` anyway, so it doesn't matter what the scope is here. But yeah I'll have another look at this, I might change them back to traditional functions - I think I'm doing this everywhere in the app though so might not be part of this PR.

meirish replied:
I like arrow functions, but yeah using `arguments` can be tricky - but this is doing what you intended so 👌

johncowen commented on ui-v2/app/routes/dc/acls/index.js:
```diff
@@ -13,6 +13,9 @@ export default Route.extend(WithAclActions, {
     },
   },
+  beforeModel: function(transition) {
+    this.replaceWith('dc.acls.tokens');
```
Good spot, ta!

johncowen commented on ui-v2/app/templates/components/tabular-collection.hbs:
```diff
@@ -1,22 +1,25 @@
+      </tr>
+  </thead>
+  {{#ember-native-scrollable tagName='tbody' content-size=_contentSize scroll-left=_scrollLeft scroll-top=_scrollTop scrollChange=(action "scrollChange") clientSizeChange=(action "clientSizeChange")}}
+      <tr></tr>
```
`ember-collection` sizes the scrollbar by using the first child in the DOM element. It sets the height of this first child to the size of the scrolled content. I don't want it sizing the first row of my table so this empty first child goes here for `ember-collection` to resize. The only reason this is in this diff is that the whitespace changed.

meirish replied:
ember-collection sounds fun 😬

johncowen commented on ui-v2/app/templates/components/token-list.hbs:
```diff
@@ -0,0 +1,27 @@
+      {{/block-slot}}
+      {{#block-slot 'row'}}
+          <td data-test-token="{{item.AccessorID}}">
+            <a href={{href-to 'dc.acls.tokens.edit' item.AccessorID}} target={{or target ''}}>{{truncate item.AccessorID 8 false}}</a>
```
I wouldn't have thought so? This is an internal link. Those attributes are to try to hide where links are coming from so you generally only put them on external links.

meirish replied:
Ah yeah it was just with the `target` that threw me off

johncowen commented on ui-v2/app/templates/dc/acls/-authorization.hbs:
```diff
@@ -0,0 +1,13 @@
+<p>
```
👍 Once template only glimmer components are available for me, I'll look to switch. Right now having to have an empty component file aswell as a template is just extra overhead. This particular one I might be even moving to my `app-view` component. A lot of the others will go when I do part 2 of my form thing.

johncowen commented on ui-v2/app/templates/dc/acls/-form.hbs:
```diff
@@ -27,9 +27,7 @@
-        <button type="submit" {{ action "create" item}} disabled={{if (or item.isPristine item.isInvalid (eq item.Name '')) 'disabled'}}>Save</button>
-{{ else }}
-        <button type="submit" {{ action "update" item}} disabled={{if item.isInvalid 'disabled'}}>Save</button>
+        <button type="submit" {{ action "create" item}} disabled={{if (or item.isPristine item.isInvalid (eq item.Name '')) 'disabled'}}>Save</button>{{else}}<button type="submit" {{ action "update" item}} disabled={{if item.isInvalid 'disabled'}}>Save</button>
```
Yeah for sure, this is actually on my list of improvements. I think this style actually came from me converting the old v1 UI 😂 . Probably a good feature PR following this, needs to be done app wide in one go.

johncowen commented on ui-v2/app/templates/dc/acls/policies/-form.hbs:
```diff
@@ -0,0 +1,46 @@
+            {{#block-slot 'dialog' as |execute cancel message|}}
+    {{#if (gt items.length 0)}}
+              {{#modal-dialog onclose=(action cancel)}}
+                  {{#block-slot 'header'}}
```
They are scoped to the component, as far as I know.

johncowen commented on ui-v2/app/templates/dc/acls/tokens/index.hbs:
```diff
@@ -0,0 +1,134 @@
+                                </li>
+{{#if (not item.Legacy)}}
+                                <li>
+                                    <a data-test-clone onclick={{action 'sendClone' item}}>Duplicate</a>
```
Still in my list to do! I'd rather do it in a separate PR, not in a ACL upgrade PR.

johncowen replied:
Was just re-reviewing this PR and remembered this. Thought I'd pop the PR I eventually did in here for reference:

https://github.com/hashicorp/consul/pull/4939

johncowen commented on ui-v2/app/templates/dc/acls/policies/-fieldsets.hbs:
```diff
@@ -0,0 +1,49 @@
+{{/if}}
+      <label class="type-text">
+        <span>Description (Optional)</span>
+        <textarea name="policy[Description]" value={{item.Description}} oninput={{action 'change'}}></textarea>
```
> This event is similar to the onchange event. The difference is that the oninput event occurs immediately after the value of an element has changed, while onchange occurs when the element loses focus.

johncowen commented:
Hey @meirish 

Thanks so much for looking at this, I've spent a good deal of time just responding to these, so I can imagine the time spent going over it.

I know it looks like a lot of work, but its not really - the grand majority of it is just copying over what's already been done. A new Adapter/Serializer/Model (partly taken from the old ACLs and moved folders), a set of new Repo/Route/Controller/Templates for both policy/tokens which follow the exact same things as the rest of the app. I think a lot of your comment above are just not being massively familiar with the history/previous PR's. I've tried to comment things in above to answer your queries, but we can go over the entire architecture again briefly before a full review.

I think the biggest decision for me is.. what do you think of the `SingleRoute` thing? This is the thing I'm very on the fence about. Anyway, we can chat through that when we chat.

Thanks again for the time with this!

John

meirish commented on ui-v2/app/routing/single.js:
```diff
@@ -0,0 +1,28 @@
+      create: create,
+      ...repo.status({
+        item: create
+          ? Promise.resolve(repo.create({ Datacenter: dc }))
```
Is there ever a chance you'd need to do findOrCreate style fetch here?

johncowen commented on ui-v2/app/templates/components/tab-nav.hbs:
```diff
@@ -1,9 +1,13 @@
+{{#if item.href }}
+  <a href="{{ item.href }}">{{ item.label }}</a>
+{{ else }}
+  <a>{{ item }}</a>
```
Just 're-seen' this. Gonna come back to this for a later date, the attribute-less <a> isn't in this PR (if you look at the removed code it was there also), but it could do with re-looking at. I vaguely remember this being something to do with being able to have a tab bar that changes without changing the URL or any state, or having a tab bar where the tabs act as links - also I think this makes the tabs more accessible in that they are automatically in the tablist when you tab through the page. I've added this to my todo to come back later.

johncowen commented:
Hey @meirish 

I've changed the base branch for this now and clean up some of the commented in code that was helping me during dev. Since we spoke yesterday I've also pulled out the storage prefixing into its own module so I can unit test it (and added unit tests).

I think the only thing I might add to this before you are about today is moving `siblings` and `closest` into my `dom` service and move everything within ACLs to use the `dom` service instead of importing directly (if I get chance, so please don't let that stop you giving a 👍 )

Other than that, I'm just looking for an approval on this now

Cheers!

meirish commented:
Oh @johncowen - was also curious if you had any response to the ember-wormhole bit I mentioned here: https://github.com/hashicorp/consul/pull/4789#pullrequestreview-164749118 (no rush on that though!)

johncowen commented:
> I was curious about the ember-wormhole removal though - did you run into any problems with it or just didn't like the use of private apis? With the direct dependency on it removed, aren't you still loading it transitively since you're using power-select? What's the plans for the future there (I think power-select now uses that maybe-in-element addon) ?

Ah I've been looking for that comment but couldn't find it, I've been looking in the wrong place (the inline comments)...anyway..

I didn't remove it as such, I just didn't use it.

From what I can see nobody wants to use ember-wormhole, everybody that used to use it is moving away from it, I've even seen the team talking about problems using it. Then of course there was the private API usage.

I looked for others and only found that ember-elsewhere, which I tried but as far as I can see you have to specify your entire template/DOM in an attribute, which wasn't ideal 🤷‍♂️ .

I've not looked/tried that maybe-in-element addon as I wasn't aware of it until you mentioned it (it didn't surface whilst looking for alternatives), but from what I can see it also uses private APIs (that may soon be made public), plus it seems to be a transpiler rather than actual javascript, which immediately made me think a little too strange for now.

In the end I took a quick pass at making my own and it only took a little while, probably shorter than trying out other things, so I went with that.

The modals will be coming out reasonably soon anyway as theres another style of UI coming down the line pretty soon.

> With the direct dependency on it removed, aren't you still loading it transitively since you're using power-select?

I 'think' even power-select are using that maybe-in-element one now, but they have a load of code in there to cope with both approaches.. (I was browsing over the power-select and simple-menu source last night)

DingoEatingFuzz commented:
@johncowen, I haven't looked at the code yet but excellent write up 👏 👏. Judging by how much you have to say about the high-level decisions after the fact, it makes me think this would have been a good candidate for an RFC before implementation.

Gonna peruse the code now 👀

DingoEatingFuzz commented on ui-v2/app/templates/components/modal-dialog.hbs:
```diff
@@ -0,0 +1,19 @@
+{{yield}}
+<input id={{name}} type="radio" name="modal" checked={{checked}} onchange={{action 'change'}} />
```
I don't follow what this input is doing at all. The state has the unhelpful name `checked` and the action has the unhelpful name `change`.

DingoEatingFuzz commented on ui-v2/app/utils/form/builder.js:
```diff
@@ -0,0 +1,156 @@
+  return changeset;
+};
+/**
+ * Form builder/Form factory (WIP)
```
Seeing WIP on the master branch is never reassuring.

johncowen replied:
No prob, removed in https://github.com/hashicorp/consul/pull/5021

DingoEatingFuzz commented on ui-v2/ember-cli-build.js:
```diff
@@ -5,7 +5,8 @@ module.exports = function(defaults) {
   const isProd = env === 'production';
-  const isProdLike = prodlike.indexOf(env) > -1;
+  // leave this in for now for when I start a proper staging env
+  // const isProdLike = prodlike.indexOf(env) > -1;
```
A couple opinions on this, take them or leave them. 1) Saying things like "when I start" in code comes off kind of controlling. Am I supposed to do a git blame to see who "I" is? Should I leave this alone and let "I" deal with it? 2) It seems easy enough to add this line back once there is a staging env rather than having an unused wart for a possible future.

johncowen replied:
No prob,I tweaked the comment in https://github.com/hashicorp/consul/pull/5021 to reword it slightly.

You don't have to, but to explain a little - if you look at the blame for this file:

https://github.com/hashicorp/consul/blame/88027a940b088d4e3c96b8dbcd9e557c8d524881/ui-v2/ember-cli-build.js

You'll see I did it during a 'Comment clean up' commit. I didn't remove this unused code on purpose, but added the comment to explain why I didn't remove it like I did with others. For me it will be useful at somepoint, maybe for others also. It's in 'node-world' so it doesn't end up in shipping code (and is a comment anyway), so I've chosen to leave it in for the moment.

We are currently looking at setting up a more permanent staging environment for the UI, so this might come in useful 🔜

DingoEatingFuzz commented on ui-v2/package.json:
```diff
@@ -34,7 +34,7 @@
   },
   "devDependencies": {
-    "@hashicorp/consul-api-double": "^1.5.3",
+    "@hashicorp/consul-api-double": "^2.0.1",
```
A major version bump? What changed? Do you PR to this repo? Do the changes get code reviewed?

johncowen replied:
> A major version bump? What changed?

Yup, major. I changed the name of one of the cookies you can use to debug/test from `CONSUL_ENABLE_ACLS` to `CONSUL_ACLS_ENABLE` to bring it in line with all of the other variables I use to debug and test the UI (they all follow `CONSUL_<MODEL_NAME>_<SETTING>` now). So, unfortunately breaking yes. I'm the only person that I know of using this so I'd say it's not a big deal in the scheme of things, but you never know, plus its habit I suppose. Breaking means major bump.

> Do you PR to this repo? Do the changes get code reviewed?

No I don't PR changes, I commit straight onto `master` and publish to `npm` to release, if there is a 'bug' I release a patch version, but generally it's reasonably well tested first considering what it is. This repo is only my 'api double' _templates_. They don't really do anything, it's only a templated version of the `consul` HTTP API. So its pretty much a tonne of JSON text files.

Also, the module isn't shipping code (in the Consul UI), it's only to make it easier for me to test using massive amounts of data.

I sometimes ask the backend team to review the code if it's for a new upcoming API that hasn't been nailed down yet to make sure I can build out the UI without the backend being built.

If you want to help out with `api-double` (the actual template renderer, not the specific `consul-api-double` templates) I'd be more than happy for you to pitch in if that's what you want to do. I've mention people helping out a few times but up until now no-one seems interested? Let me know, happy to walk you through.

johncowen replied:
I just noticed this whilst I was searching for the 'DDAU alternative' PR, thought it gives some good context `api-double` vs `consul-api-double` so posting here for reference:

https://github.com/hashicorp/consul/pull/4190#discussion_r193176956

johncowen commented on ui-v2/app/components/modal-dialog.js:
```diff
@@ -0,0 +1,119 @@
+import { get, set } from '@ember/object';
+import { inject as service } from '@ember/service';
+import Component from 'consul-ui/components/dom-buffer';
```
I changed it to `DomBufferComponent` so it's massively clear what it is. See https://github.com/hashicorp/consul/pull/5021

I also changed it in other places such as my `ember-collection` extended `TabularCollection` to be consistent.

johncowen commented on ui-v2/app/components/modal-dialog.js:
```diff
@@ -0,0 +1,119 @@
+    set(this, 'checked', true);
+    if (get(this, 'height') === null) {
+      if (this.element) {
+        const dialogPanel = get(this, 'dom').element('[role="dialog"] > div > div', this.element);
```
Thanks for the suggestion here, I can partly see where you are coming from, but...

I suppose this is similar to the conversation we had around CSS selectors and HTML/template structure.

Usually if I need to change one, it is absolutely no problem to change the other. Indeed in this case other frameworks like React and Vue etc let you, even encourage you, to keep your template in the exact same file as your javascript behavior (that's not something I'm keen on myself).

As far as I'm aware with the ember module unification effort, all of these files will live in the same place, coupling them even more and making them even easier to make changes like this.

> I'd feel a little better if it were an id

`id`s should be unique. With this component it's probably not the case, but if I were to add a component to the same page twice, and the component template used an `id` in it, it could produce one of those hard to track down bugs. So as a general rule I would almost never put an `id` in a component.

johncowen commented on ui-v2/app/controllers/dc/acls/policies/edit.js:
```diff
@@ -0,0 +1,45 @@
+        return prev;
+      }, model)
+    );
+    set(this, 'isScoped', get(model.item, 'Datacenters.length') > 0);
```
> I liken it to overriding an Array or String prototype. Sure Array#push is a public method but anyone familiar with JavaScript has a preconceived idea of what it does

I'd definitely not recommend altering any native objects in any way without extending them first. I think we've spoken about this before here https://github.com/hashicorp/nomad/pull/4635#discussion_r217655431

I've just looked again at the file, aren't you still overriding Array instance methods here? 

https://github.com/hashicorp/nomad/blob/adc05976c2c7f97d46c214518d448b0b13e377d6/ui/app/utils/classes/rolling-array.js

It's not the prototype method, but you couldn't you still do an

```javascript
array.constructor === Array ? array.push(thing) : somethingElse
```

 ...and end up with something you didn't expect?

In my case I'm _extending_ `Controller` first, so I'm essentially adding functionality to a base `Controller` to give me a specific `PoliciesController` which is my own custom class. `setProperties` is public and not final in anyway (AFAIK) so usually that would mean 'fair game', as long as it does what you expect aswell as the extra functionality you add (which it does by calling `super`). Additionally what my custom `setProperties` does is to 'proxy' certain variables you pass it (which I've mentioned here https://github.com/hashicorp/consul/pull/4789/files/88027a940b088d4e3c96b8dbcd9e557c8d524881#r225548050). I also left a comment in to explain this slightly:

https://github.com/hashicorp/consul/blob/88027a940b088d4e3c96b8dbcd9e557c8d524881/ui-v2/app/controllers/dc/acls/policies/edit.js#L13

Changesets are just proxy objects to the underlying data. So it just 'decorates' the data with extra functionality.

> Something like controller.setProperties(controller.syncFormData({ prop, er, ties })) would be more obvious upon reading.

One thing I should explain here, is that I went to great lengths to keep the `forms` in the View/ViewController layer, and out of the Routes. They are a view concern, so this is where they should live.

I can appreciate your overall point a little bit and I did think of other options and I've thought of others to try to address this somehow. The answer to the problem here is to find a hook to use in the View layer immediately before the View is rendered. I've considered `setProperties` to be this hook as the View is re-rendered immediately after that.

Other things I can think of:

1. Use a computed property: 
```javascript
actualItem: computed('item', function(){
  return this.form.setData()
})
```
This now means that `item` in my template is no longer the `item` that I want. It's the unproxied `item`. To me this is surprising as I'm passing it into the `View` as `item`, if I look at the templates it leaves me thinking "what's `actualItem`, what happened to `item` from my Route?" Plus I'd have to do this for every property I want to proxy, which would get tedious very quickly. Right now I can just loop over the props in `setProperties` and proxy the object there.

2. Use computed setter/getters:
```javascript
item: computed('item', function(){
  {
    set: function(){},
    get: function(){}
  }
})
```

As far as I know this means having yet another property on the Controller to save the actual value of `item`, I'd also have to do this for every property (see above).

3. Use a more generically named controller 'setup' function, and remember to call it everytime I call `setProperties`.
```javascript
export Route.extend({
  setupController(controller, model) {
   
    controller.setProperties({
      some: 'things',
      go: here,
    });
    if(typeof controller.setup === 'function') {
      controller.setup();
    }
    return this._super(...arguments);
  },
});
```
`controller.setup` would mutate the `item` property into the proxy object. But it's named something generic enough to make sense. I'm actually surprised that I can't find any default `setup/reset`-like methods for `Controller` in `ember`.

Option 3 is my favourite at a pinch, but there is always the danger that you'd forget to call `setup` somewhere, and therefore produce some hard to spot bugs.

There's even a plugin which attempts to make this less error prone for you:

https://github.com/kellyselden/ember-controller-lifecycle

The only problem with this is, there is _still_ a danger that you might call `setProperties` somewhere else and your `controller.setup()` is never called, which would lead to unexpected bugs.

Do you have any further ideas following the above info?

(Apologies if this comment is a bit lengthy, I want to try to make sure follow my thinking here)

johncowen replied:
Sorry had another idea for this, thought I'd pop it in here incase.

> Option 3 is my favourite at a pinch, but there is always the danger that you'd forget to call setup somewhere, and therefore produce some hard to spot bugs.

Option 4:

If I changed things so that I stopped calling `setProperties`, but instead called a method on the `Controller` that _then_ called `setProperties`. So something like:

```javascript
// ReopenedController using a similar approach to ember-controller-lifecycle
newMethodNameThatsGenericEnough: function(model) {
  const proxiedModel = this.doExtraStuffWithModel();
  this.setProperties(proxiedModel);
}

// SomeRoute
setupController: function(controller, model) {
  this._super(...arguments);
  this.newMethodNameThatsGenericEnough(model);
}
```

..and now when I have anywhere in my codebase where I was previously calling `setProperties` and there was the danger of forgetting to then call a `setup`, I just change it to call `newMethodNameThatsGenericEnough` instead.

I don't think this method should be called `setup` though as I also call `setProperties` when I'm not setting things up, so it wouldn't make sense there. Will think a bit more on a good name. But I think this addresses my problem and satisfies your concerns about overwriting the `setProperties` method, so win win.

The only tiny thing is, this isn't 'conventional `ember`' we've added a new undocumented method - but I'm ok with that if you are?

johncowen replied:
Related: https://github.com/hashicorp/consul/pull/5056

johncowen commented on ui-v2/app/styles/components/index.scss:
```diff
@@ -0,0 +1,31 @@
+@import './feedback-dialog';
+@import './modal-dialog';
+@import './notice';
+@import './with-tooltip';
```
It was just that the `app.css` file was getting a bit too full of component file imports 😬 - my app is getting big enough now that I thought it was probably a good time to give them their own `index` file.

johncowen commented:
Hey @DingoEatingFuzz,

Thanks for the review on this, sorry for taking so long to come back to you on your comments. I've tried to explain a little more so you can see my thinking on some of the points, and I've made a few of changes in https://github.com/hashicorp/consul/pull/5021 based on your review.

Thanks
